infrastructure/containerd/image_service_integration_test.go

```diff
@@ -0,0 +1,291 @@
+package containerd_test
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+
+	ctr "github.com/containerd/containerd"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/snapshots"
+	. "github.com/onsi/gomega"
+
+	"github.com/weaveworks/flintlock/core/models"
+	"github.com/weaveworks/flintlock/core/ports"
+	"github.com/weaveworks/flintlock/infrastructure/containerd"
+)
+
+const (
+	testImageVolume    = "docker.io/library/alpine:3.14.1"
+	testImageKernel    = "docker.io/linuxkit/kernel:5.4.129"
+	testSnapshotter    = "devmapper"
+	testOwnerNamespace = "int_ns"
+	testOwnerUsageID   = "vol1"
+	testOwnerName      = "imageservice-get-test"
+	testContainerdNS   = "flintlock_test_ctr"
+)
+
+// What this test does?
+//
+// == Preface
+//
+// Prepare ImageService and a Namespace context. All images and snapshots will
+// live under the same namespace.
+//
+// We do some prudent cleanup steps in case the test was aborted. Because it's
+// an integration test with an external service, if we don't clean up after
+// ourselves, the next test run will most likely fail. In theory, releasing the
+// leases should be enough, but that's heavily depend on how containerd tells
+// us how it works, so it's easier to delete everything we intended to create.
+//
+// == Chapter I: Pull and mount
+//
+// Using ImageService, we tries to Pull an image.
+//
+// We don't have separate Mount function (yet), we try to PullAndMount, with
+// this, we can be sure ImageService can pull and image from a repository and
+// can mount it.
+//
+// To avoid false positive result, we try to PullAndMount an image, we know
+// it's definitely not there, and we expect it to be failed.  If it does not
+// fail, something fishy is swimming under the hood.
+//
+// After PullAndMount, the real image should be mounted, but the fake one
+// shouldn't be. At the end we should have only one image available locally.
+//
+// == Chapter II: Snapshots
+//
+// As we already pull and mounted an image, we expect one snapshot to be there
+// with the name constructed from the VM namespace, the VM name, and the volume
+// usage type.
+// At the same time, we expect one lease to be there with similar naming
+// structure except the volume usage type we have one lease for all resources
+// in containerd.
+//
+// == Chapter III: Kernel
+//
+// We do the same checks for kernel. The kernel and volume works the same way,
+// and the real reason why we test it here because the usage type is different,
+// and we should be able to pull and mount them.
+//
+// == Chapter IV: Delete
+//
+// We should be able to delete both the volume and the kernel image. After we
+// delete them, we should see zero images in containerd.
```

Comment:
can we re-word this to explain we are testing?

from what i can see, the deletion is not us so we don't really care much about it. the test from what i can see is actually on `Exists`

honestly the whole delete bit could be moved to cleanup as it doesn't have anything to do with us and we already tested Exist earlier.

Reply:
actually, where does our code call various Deletes on the containerd services? I expected to see it around here but can't 🤔

Reply:
https://github.com/weaveworks/flintlock/pull/269/files/113cb403c180621e966325abae9d7776bf09e65a#diff-d96f2d206636039c31d85ae360ad7739ae7487bd2c5847daffa4db30446d0135R215-R219

Reply:
So it's:
1. delete images
2. check if they are there (one by one)
3. check List, it should be zero.
```
	//
	// Chapter IV: Delete
	//

	err = client.ImageService().Delete(namespaceCtx, getTestKernelImage())
	Expect(err).NotTo(HaveOccurred())

	err = client.ImageService().Delete(namespaceCtx, getTestVolumeImage())
	Expect(err).NotTo(HaveOccurred())

	exists, err = imageSvc.Exists(ctx, &ports.ImageSpec{
		ImageName: getTestKernelImage(),
		Owner:     testOwnerUsageID,
	})
	Expect(err).ToNot(HaveOccurred())
	Expect(exists).To(BeFalse())

	images, err := client.ImageService().List(namespaceCtx)
	Expect(err).NotTo(HaveOccurred())
	Expect(len(images)).To(BeZero())
```

Reply:
`client.ImageService().Delete` is not our code right? we are just calling containerd directly?

Reply:
where do we actually call delete on things? like in the code not in tests? i can't find where flintlock deleted containerd things on mvm teardown. or wait do we not do that yet? is that one of the things we left for later? yes i am remembering now we have a thing for it. nvm as you were

Reply:
The I would say just remove that part and keep the defer cleanup section as it is. I see no reason to test if it has any errors or not, not out code and the connection should be fine otherwise the rest of the test would fail anyway.

Reply:
👍  sounds good

and yeh i expected that we would have more of a delegation "shim" around containerd (and its various services) rather than just having the client and using it directly. maybe we can move towards that eventually with some refactoring, as it could lead us to a cleaner pattern which would be easier to test in-depth